kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java

```diff
@@ -0,0 +1,225 @@
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.stream.annotation.Bindings;
+import org.springframework.cloud.stream.annotation.EnableBinding;
+import org.springframework.cloud.stream.messaging.Source;
+import org.springframework.context.annotation.Bean;
+import org.springframework.integration.kafka.core.Configuration;
+import org.springframework.integration.kafka.core.ConnectionFactory;
+import org.springframework.integration.kafka.core.DefaultConnectionFactory;
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.integration.kafka.core.ZookeeperConfiguration;
+import org.springframework.integration.kafka.inbound.KafkaMessageDrivenChannelAdapter;
+import org.springframework.integration.kafka.listener.KafkaMessageListenerContainer;
+import org.springframework.integration.kafka.listener.KafkaNativeOffsetManager;
+import org.springframework.integration.kafka.listener.OffsetManager;
+import org.springframework.integration.kafka.support.ZookeeperConnect;
+import org.springframework.retry.RetryCallback;
+import org.springframework.retry.RetryContext;
+import org.springframework.retry.backoff.ExponentialBackOffPolicy;
+import org.springframework.retry.policy.SimpleRetryPolicy;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.StringUtils;
+
+import kafka.serializer.Decoder;
+import kafka.serializer.DefaultDecoder;
+import kafka.utils.VerifiableProperties;
+
+/**
+ * Source module that receives data from Kafka.
+ *
+ * @author Soby Chacko
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties(KafkaConfigurationProperties.class)
+public class KafkaSource {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private static final int METADATA_VERIFICATION_RETRY_ATTEMPTS = 10;
+
+    private static final double METADATA_VERIFICATION_RETRY_BACKOFF_MULTIPLIER = 2;
+
+    private static final int METADATA_VERIFICATION_RETRY_INITIAL_INTERVAL = 100;
+
+    private static final int METADATA_VERIFICATION_MAX_INTERVAL = 1000;
+
+    @Autowired
+    @Bindings(KafkaSource.class)
+    private Source source;
+
+    @Autowired
+    private KafkaConfigurationProperties properties;
+
+    @Bean
+    public KafkaMessageListenerContainer container() {
+        Partition[] partitions = getPartitions();
+
+        KafkaMessageListenerContainer container = ObjectUtils.isEmpty(partitions) ?
+                new KafkaMessageListenerContainer(kafkaConnectionFactory(), properties.getTopics()) :
+                new KafkaMessageListenerContainer(kafkaConnectionFactory(), partitions);
+
+        // if we have less target partitions than target concurrency, adjust accordingly
+        container.setConcurrency(Math.min(properties.getConcurrency(), partitions.length));
```

Comment:
OK. Never mind.
I see now the reason. `ConcurrentMessageListenerDispatcher` has a code like:
```
Assert.isTrue(consumers <= partitions.size(),
				"The number of consumers must be smaller or equal to the number of partitions");
........
this.consumers = Math.min(partitions.size(), consumers);
```
So, we don't reach that `Math.min()` if we provide more consumers than partitions.

I guess that `Assert` should be removed in the SIK. The  `Math.min()` is a good compromise do not disturb end-user.